crates/via-router/src/routes.rs

```diff
@@ -0,0 +1,162 @@
+/// internal state of the route store consistent.
+pub struct RouteEntry<'a, T> {
+    /// The key of the node that we are currently working with.
+    key: usize,
+
+    /// A mutable reference to the route store that contains the node.
+    store: &'a mut RouteStore<T>,
+}
+
+/// A container type used to improve the cache locality of nodes and routes in
+/// the route tree.
+pub struct RouteStore<T> {
+    /// A collection of nodes that represent the path segments of a route.
+    nodes: Vec<Node<T>>,
+}
+
+impl<T> Node<T> {
+    pub fn new(pattern: Pattern) -> Self {
+        Self {
+            pattern,
+            entries: None,
+            route: None,
+        }
+    }
+
+    /// Returns an iterator that yields the indices of the nodes that are
+    /// reachable from `self`.
+    pub fn entries(&self) -> slice::Iter<usize> {
+        match &self.entries {
+            Some(entries) => entries.iter(),
+            None => [].iter(),
+        }
+    }
+
+    /// Returns an optional reference to the name of the dynamic parameter
+    /// associated with the node. The returned value will be `None` if the
+    /// node has a `Root` or `Static` pattern.
+    pub fn param(&self) -> Option<&'static str> {
+        match &self.pattern {
+            Pattern::CatchAll(param) | Pattern::Dynamic(param) => Some(param),
+            _ => None,
+        }
+    }
+
+    /// Returns an optional reference to the route associated with the node.
+    pub fn route(&self) -> Option<&T> {
+        self.route.as_ref()
+    }
+
+    /// Returns a mutable reference to the `route` field of the node.
+    pub fn route_mut(&mut self) -> &mut Option<T> {
+        &mut self.route
+    }
+}
+
+impl<T> Node<T> {
+    /// Pushes a new key into the entries of the node and return it.
+    fn push(&mut self, key: usize) -> usize {
+        self.entries.get_or_insert_with(Vec::new).push(key);
+        key
+    }
+}
+
+/// Like `Option::expect`, but in B minor.
+macro_rules! unwrap_node {
+    (
+        // The option that contains the `&Node<_>` or `&mut Node<_>` that we'll
+        // attempt to unwrap.
+        $option:expr,
+        // The key that was used to look up the node contained in `$option` from
+        // the route store.
+        $at:expr
+    ) => {
+        match $option {
+            Some(node) => node,
+            None => {
+                // This should never happen.
+                //
+                // The router is like the Hotel California. You can check out any
+                // time you like, but you can never leaf.
+                //
+                // If you see this error, you've probably found a bug in the way
+                // that routes are added to the route store or the way the key
+                // is assigned to nodes.
+                if cfg!(debug_assertions) {
+                    panic!("unknown key: {}", $at);
+                }
+
+                panic!("unknown key");
```

Comment:
I was considering coming up with a non-panicking version of this business logic with hooks to tracing if necessary. However, considering that the only case where this would occur is if the route store's memory is corrupt. In the case of corrupt memory, the safest thing to do is to panic. While this is counterintuitive to the idea of maximum uptime and high availability, it's a choice we make to prefer downtime over sending the wrong or incomplete message. As always, I'm open to feedback as to whether or not my thinking is sound here.